Kernel/SubstHelper.hpp

```diff
@@ -26,6 +26,45 @@ namespace Kernel {
 
 using namespace Lib;
 
+struct SubstApplicator {
+  virtual ~SubstApplicator() = default;
+  virtual TermList operator()(unsigned v) const = 0;
+};
+
+/**
+ * Term with a substitution applied to it lazily.
+ */
+struct AppliedTerm
+{
+  TermList term;
+  bool aboveVar;
+  const SubstApplicator* applicator;
+
+  AppliedTerm(TermList t) : term(t), aboveVar(false), applicator(nullptr) {}
+  /**
+   * The substitution is only applied to the term @b t via @b applicator if @b t is a variable
+   * and @b aboveVar is true (i.e. we are still above the substitution). @b applicator can be
+   * null is @b aboveVar is false.
+   */
+  AppliedTerm(TermList t, const SubstApplicator* applicator, bool aboveVar)
+    : term(aboveVar && t.isVar() ? (*applicator)(t.var()) : t),
+      aboveVar(aboveVar && t.isVar() ? false : aboveVar), applicator(applicator) {}
+
+  AppliedTerm(TermList t, AppliedTerm parent)
+    : term(parent.aboveVar && t.isVar() ? (*parent.applicator)(t.var()) : t),
+      aboveVar(parent.aboveVar && t.isVar() ? false : parent.aboveVar), applicator(parent.applicator) {}
+
+  /**
+   * Only allow comparisons if we can guarantee that the terms are the same.
+   */
+  bool operator==(AppliedTerm) const = delete;
+  bool operator!=(AppliedTerm) const = delete;
```

Comment:
Out of interest: What is this needed for?
I guess something with subtyping ...? I don't really use subtyping so I'm not sure what it's supposed to prevent.

Reply:
This is done so that the default generated operators are not used accidentally, thinking that they do a proper (and expensive) deep equals check when they in fact don't. We use the more suggestively named operator below.

Reply:
Ah right i see! Yeah i like having the explicit shallow equals operation!